src/test.rs

```
@@ -0,0 +1,89 @@
+}
+
+impl TestDefinition {
+    /// Parse the user-defined input matrix into an iterator of ordered maps of
+    /// input names to values.
+    ///
+    /// Each map represents a set of input keys whose values should be iterated
+    /// through together. The trivial case is a single input key with a set
+    /// of possible values. Groups of inputs that should be iterated through
+    /// together are designated by a YAML map key starting with `$`.
+    pub fn parse_inputs(&self) -> impl Iterator<Item = InputMapping> {
+        self.inputs.iter().map(|(key, val)| {
+            let Value::String(k) = key else {
+                panic!("expected string, got `{key:?}`");
+            };
+            if k.starts_with('$') {
+                let Value::Mapping(map) = val else {
+                    panic!("expected mapping, got `{val:?}`");
+                };
+                let mut new_map = IndexMap::new();
+                for (nested_key, nested_val) in map {
+                    let Value::String(k) = nested_key else {
+                        panic!("expected string, got `{nested_key:?}`");
+                    };
+                    let Value::Sequence(vals) = nested_val else {
+                        panic!("expected sequence, got `{nested_val:?}`");
+                    };
+                    new_map.insert(k.to_string(), vals.to_vec());
+                }
+                new_map
+            } else {
+                let Value::Sequence(vals) = val else {
+                    panic!("expected sequence, got `{val:?}`");
+                };
+                IndexMap::from_iter(vec![(k.to_string(), vals.to_vec())])
+            }
+        })
+    }
+
+    /// Parse the defined assertions into an ordered map.
+    pub fn parse_assertions(&self) -> IndexMap<String, Value> {
+        self.assertions
+            .iter()
+            .filter_map(|(k, v)| {
+                if !k.is_string() {
+                    warn!("skipping non-string key: `{:?}`", k);
```

This feels like a "return an error" condition to me rather than a logged warning.